### ISSUE TITLE
`conn->cselect_bits` is cleared after one easy handle read the socket in a shared `data->conn` when multi handle multiplex a connection

### ISSUE THREAD

**Issue by MrDwZ** [curl/curl#7977]
### I did this
1) I added multiple easy handle with the same url supporting HTTP/2 to a multi-handle and driving by an eventbase, run in a loop and set `CURLMOPT_PIPELINING` with `CURLPIPE_MULTIPLEX` to make sure connection get reused.
2) Adding a debug line in [transfer.c](https://github.com/curl/curl/blob/master/lib/transfer.c#L1225) as the following
```
...
infof(
    data, "k->keepon & KEEP_RECV : %d, select_res & CURL_CSELECT_IN %d %p",
    k->keepon & KEEP_RECV, select_res & CURL_CSELECT_IN, data);
  if((k->keepon & KEEP_RECV) && (select_res & CURL_CSELECT_IN)) {
    result = readwrite_data(data, conn, k, &didwhat, done, comeback);
    if(result || *done)
      return result;
  }
```
and [multi.c](https://github.com/curl/curl/blob/8e701cc978430d638517d025fade2b0dc1ae9b73/lib/multi.c#L3027)
```
if(data->conn && !(data->conn->handler->flags & PROTOPT_DIRLOCK)) {
          infof(data, "%p with conn: %p set ev_bitmask %d ", data, data->conn, ev_bitmask);
          /* set socket event bitmask if they're not locked */
          data->conn->cselect_bits = ev_bitmask;
}
```
3) Drive with a event base using `curl_multi_socket_action`

I observed that these easy handles will share the same `conn` object and has `ev_bitmask` set for this connection, below I added 3 different easy handles with the same URL and the eventbase fire read event and call curl_multi_socket_action.
```
== Info: 0x62200019b100 with conn: 0x61c000040080 set ev_bitmask 1 
== Info: 0x622000199900 with conn: 0x61c000040080 set ev_bitmask 1 
== Info: 0x62200019c900 with conn: 0x61c000040080 set ev_bitmask 1 
```

When `curl_multi_socket_action` is called with a fired event, the event doesn't fan out to all the easy handle that multiplex the same conn (socket), thus only 1 easy handle read the socket and finish the HTTP session, other easy handles finish upon a expiration of either connect timeout or total timeout, because the socket doesn't fire any additional event.

```
== Info: k->keepon & KEEP_RECV : 1, select_res & CURL_CSELECT_IN 1 0x62200019b100
== Info: k->keepon & KEEP_RECV : 1, select_res & CURL_CSELECT_IN 0 0x622000199900
== Info: k->keepon & KEEP_RECV : 1, select_res & CURL_CSELECT_IN 0 0x62200019c900
== Info: STATE: PERFORMING => DONE handle 0x62200019b100; line 2376 (connection #0)
...after connect timeout
== Info: k->keepon & KEEP_RECV : 1, select_res & CURL_CSELECT_IN 1 0x622000199900
== Info: k->keepon & KEEP_RECV : 1, select_res & CURL_CSELECT_IN 1 0x62200019c900
== Info: STATE: PERFORMING => DONE handle 0x62200019c900; line 2376 (connection #0)
```
Seems [`Curl_readwrite`](https://github.com/curl/curl/blob/da2598f823e95020defd54c6fdf3c82331ef12b3/lib/transfer.c#L1185) will clear the cselect_bits after a single easy handle read from the conn object, thus all the other easy handles ignore the event and not reading from the conn object.

This issue does not appear when CURLMOPT_PIPELINING is set to CURLPIPE_NOTHING in my application. I tried to work on  the code in transfer.c to not reset cselect_bits, but `Curl_readwrite` returning CURLE_AGAIN sometime also causes the similar symptom.

### I expected the following
All the easy handle finished when the first ev_bitmask is set.

### curl/libcurl version
7.78.0

### operating system
Red Hat Enterprise Linux 8, kernel version Linux version 5.2.9

**Comment by bagder** [comment]
1. Any chance you can provide a small stand-alone example that reproduces this problem?
2. We've fixed HTTP/2 even after 7.78.0, so I recommend using the 7.79.1 version when debugging this

**Comment by MrDwZ** [comment]
@bagder Thank you for the quick reply.

> 1. Any chance you can provide a small stand-alone example that reproduces this problem?

I'll try, the implementation in our codebase is somehow pretty complicated with our Folly C++ library and C++20 coro. I'll see if I can re-use some simpler example to repro but in the worst case I'll sanitize the code we use and upload, may take a while. 

> 2. We've fixed HTTP/2 even after 7.78.0, so I recommend using the 7.79.1 version when debugging this

Just see 7.8.0 is released, I'll play with it!

**Comment by bagder** [comment]
If you cannot reproduce the problem with a stand-alone example code, I'm afraid we cannot do a lot of things. Also, if you fail to reproduce it with a smaller sample after all this time, chances are the problem is not (only) in curl...
